dockerfile-captain.release

```diff
@@ -1,18 +1,32 @@
-FROM node:18-alpine
-RUN apk update && apk upgrade --no-cache && apk add --update --no-cache make gcc g++ git curl openssl openssh
-
-RUN mkdir -p /usr/src/app
+# Build Source Into JS
+FROM node:18-alpine as builder
+RUN apk add --update --no-cache make gcc g++ git curl openssl openssh
 WORKDIR /usr/src/app
-
-
-# Build backend code
-
-COPY . /usr/src/app
-
+COPY . ./
 RUN npm ci && \
      npm cache clean --force && \
      npm run build 
 
+# Install only production dependencies
+FROM node:18-alpine as deps
+WORKDIR /usr/src/app
+COPY packag*.json ./
+RUN npm ci --omit=dev && \
+     npm cache clean --force
+
+# Final Image in Production
+FROM node:18-alpine as runner
+RUN apk add --update --no-cache make gcc g++ git curl openssl openssh
+WORKDIR /usr/src/app
+
+COPY --from=builder /usr/src/app/built ./built
+COPY --from=deps /usr/src/app/node_modules ./node_modules
+COPY --from=builder /usr/src/app/dist-frontend ./dist-frontend
+COPY --from=builder /usr/src/app/public ./public
+COPY --from=builder /usr/src/app/template ./template
+COPY --from=builder /usr/src/app/dockerfiles ./dockerfiles
+COPY --from=builder /usr/src/app/package.json /usr/src/app/edge-override.json ./
```

Comment:
This approach of listing out all files/dirs except node_modules is quite error prone. Instead, simply just remove the node_modules from the builder and copy the rest. Then copy from `deps` and override `node_modules` directory.

Reply:
By default, `builder` copy every files and folders from repository which are not necessary to run application. So instead, copy only those files and folders necessary to run an application.
If you want to copy the rest from `builder` except `node_modules`, you might have to add some extra lines in `.dockerignore` file. But doing so might hamper in dockerfile of `dev` and `debug` evironment.

Using my code commit, we do not have to worry about `dev` and `debug` dockerfiles

Reply:
I know that. But this is over optimizing. The main culprit here is node_modules. Other directories might be added and removed in the future and this adds to the maintenance of the application.

Reply:
I knew it! Just realized this commit as is introduces a bug. the `dev-scripts` directory is needed for Pro users if they want to [clear their OTP](https://caprover.com/docs/troubleshooting-pro.html)

```
....
docker run -it --rm -v /captain:/captain  caprover/caprover /bin/sh -c "wget https://raw.githubusercontent.com/caprover/caprover/master/dev-scripts/clear-pro-config.js ; node clear-pro-config.js ;" && \
```

Reply:
Can you mention, which others `dev-scripts` are useful so I can add only those scripts in `dev-scripts` folder.
Yes, you are correct . The main culprit is `node_modules`, but adding files and folders like `*.md`, `.git`, `.github` and even the `src` increase size of docker images. So, we must always try to keep  a minimum filesize as possible as we can.

Reply:
`.git` can be added to `.dockerignore` if it's not already. Over optimizing is the root of all evil! There is much more to software than perfection.

Reply:
> We must always try to keep a minimum filesize as possible as we can.

This is incorrect! We must always try to keep a minimum filesize without creating risks to bugs / without adding maintenance overheads, without etc etc.

Reply:
> This is incorrect! We must always try to keep a minimum filesize without creating risks to bugs / without adding maintenance overheads, without etc etc.

Yes you are correct. But, If you know what is required to run an application, will you not select only those files and folder. Yes we can keep whole `dev-scripts` folder inside `Dockerfile` if you want. But why will we need `src` folder since everything is running from compiled js code from `built` folder.

All I am saying is copy only those files and folder which are required to an application.

Plus I will add whole `dev-scripts` inside Dockerfile if you want.

Reply:
The main purpose of this PR is just to let team know that there is a part for improvement. I believe I have fulflled that.

Reply:
> I knew it! Just realized this commit as is introduces a bug. the `dev-scripts` directory is needed for Pro users if they want to [clear their OTP](https://caprover.com/docs/troubleshooting-pro.html)
> 
> ```
> ....
> docker run -it --rm -v /captain:/captain  caprover/caprover /bin/sh -c "wget https://raw.githubusercontent.com/caprover/caprover/master/dev-scripts/clear-pro-config.js ; node clear-pro-config.js ;" && \
> ```

Hey @githubsaturn, I have added dev-scripts in Dockerfile for both release and edge in commit [d2dc9ff](https://github.com/caprover/caprover/pull/2009/commits/d2dc9ff514b4a5244c94c9ee32b04971e6aeeafb).

If you still think something needs to be added, just let me know.

Reply:
@githubsaturn Any update on it?

Reply:
I am waiting for a new commit that uses exclusion logic to add the folders - as I mentioned above, inclusion logic is not scalable and error prone. Even if we include `src`, it's just a few hundreds of kb - which is astronomically small compared to the size of the image. As a design principal, CapRover does prefer maintainability and stability over over-optimization.